### PR TITLE
128bit hash-functions accepting arbitrary list of arguments

### LINKED ISSUE #28774: 128bit hash-functions accepting arbitrary list of arguments
```
select murmurHash3_64(1,2,3);

┌─murmurHash3_64(1, 2, 3)─┐
│     8163029322371165472 │
└─────────────────────────┘

SELECT sipHash64(1, 2, 3)

┌───sipHash64(1, 2, 3)─┐
│ 12940785793559895259 │
└──────────────────────┘
```

But:

```
SELECT murmurHash3_128(1, 2, 3)

Received exception from server (version 21.8.5):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Number of arguments for function murmurHash3_128 doesn't match: passed 3, should be 1: While processing murmurHash3_128(1, 2, 3). 


SELECT sipHash128(1, 2, 3)

Received exception from server (version 21.8.5):
Code: 42. DB::Exception: Received from localhost:9000. DB::Exception: Number of arguments for function sipHash128 doesn't match: passed 3, should be 1: While processing sipHash128(1, 2, 3). 
```

### PR DESCRIPTION
**Pull request by nicelulu**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This closes https://github.com/ClickHouse/ClickHouse/issues/28774.

### CODE REVIEW

**Comment by nicelulu** [comment]
https://clickhouse-builds.s3.yandex.net/28965/a843c5458896dd8be8743da4ab6dda2cc2f6bd10/clickhouse_special_build_check/build_1631527557.out.log

build error

**Comment by novikd** [comment]
@Mergifyio update

File: src/Functions/FunctionsHashing.h [hunk 1]
```diff
@@ -231,12 +231,23 @@ struct SipHash64Impl
...
+    {
+        UInt128 hashes[] = {h1, h2};
+        return apply(reinterpret_cast<const char *>(hashes), 32);
+    }
+
+    static UInt128 apply(const char * data, const size_t size)
     {
-        sipHash128(begin, size, reinterpret_cast<char*>(out_char_data));
+        char bytes[16];
+        sipHash128(data, size, bytes);
```
Comment (novikd) [inline src/Functions/FunctionsHashing.h:246]: Let's make get rid of `bytes` variable. Change `sipHash128` return type to `UInt128`.

File: src/Functions/FunctionsHashing.h [hunk 2]
```diff
@@ -354,12 +365,22 @@ struct MurmurHash3Impl64
...
     static constexpr auto name = "murmurHash3_128";
-    enum { length = 16 };
 
-    static void apply(const char * begin, const size_t size, unsigned char * out_char_data)
+    using ReturnType = UInt128;
+
+    static UInt128 combineHashes(UInt128 h1, UInt128 h2)
     {
-        MurmurHash3_x64_128(begin, size, 0, out_char_data);
+        return apply(reinterpret_cast<const char *>(&h1), 16) ^ h2;
```
Comment (novikd) [inline src/Functions/FunctionsHashing.h:373]: Hm..that doesn't look like something similar to 64-bit version.

Comment (nicelulu) [reply src/Functions/FunctionsHashing.h:373]: Thank you for your review. Do you have any suggestions? IntHash64Impl is used in the 64-bit version.

File: tests/queries/0_stateless/00678_murmurhash.sql [hunk 3]
```diff
@@ -35,3 +35,9 @@ SELECT gccMurmurHash(1);
 SELECT hex(murmurHash3_128('foo'));
 SELECT hex(murmurHash3_128('\x01'));
 
+SELECT hex(murmurHash3_128('foo', 'foo'));
+SELECT hex(murmurHash3_128('foo', 'foo', 'foo'));
+
+SELECT hex(sipHash128('foo'));
```
[the remark this answers is not in the document]
Comment (nicelulu) [reply tests/queries/0_stateless/00678_murmurhash.sql]: done.

File: src/Functions/FunctionsHashing.h (hunk 2 above)
Comment (novikd) [reply src/Functions/FunctionsHashing.h:371]: I think this should be implemented as `combine(x, y) = someHashFunction(x || y)` where `||` is "concatenation".
And probably we should change 32 and 64 bit version the same way.

Comment (nicelulu) [reply src/Functions/FunctionsHashing.h]: done.

**Review by novikd — APPROVED** [review]
LGTM. I had a conversation with the team and we decided that it's better not to modify 32 and 64 bit versions because somebody can rely on hash value. Could you please revert this changes?

**Comment by nicelulu** [comment]
> I had a conversation with the team and we decided that it's better not to modify 32 and 64 bit versions because somebody can rely on hash value.

Okay, for the 32-bit and 64-bit hash functions, I only modified the code of sipHash64, but its implementation logic has not been modified.

**Comment by ka1bi4** [comment]
Internal documentation ticket: DOCSUP-18083

File: src/Functions/FunctionsHashing.h [hunk 4]
```diff
@@ -1115,6 +1145,13 @@ class FunctionAnyHash : public IFunction
         for (const auto & col : arguments)
             executeForArgument(col.type.get(), col.column.get(), vec_to, is_first_argument);
 
+        if constexpr (std::is_same_v<ToType, UInt128>) /// backward-compatible
+        {
+            auto col_to_fixed_string = ColumnFixedString::create(sizeof(UInt128));
+            col_to_fixed_string->getChars() = std::move(*reinterpret_cast<ColumnFixedString::Chars *>(&col_to->getData()));
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsHashing.h:1151]: This does not work.

**Pull request merged**